### PR TITLE
[For testing] IPC4 timeout fixes (combo of PR6243 and PR6239)

### PR DESCRIPTION
**Pull request by kv2019i**
A collection of multiple IPC4 timeout fixes. This PR is for testing to share the fixes and to get testing widers results. I'll probably submit these as separate PRs as they are independent fixes to different parts of the codebase.

**This PR is for TESTING.**
The more mature fixses are submitted for review separate at https://github.com/thesofproject/sof/pull/6243
I will close this PR once testing is completed.

### CODE REVIEW

File: src/ipc/ipc4/handler.c [hunk 1]
```diff
@@ -986,6 +986,7 @@ void ipc_send_buffer_status_notify(void)
 	msg_notify.header = SOF_IPC4_NOTIF_HEADER(SOF_IPC4_NOTIFY_LOG_BUFFER_STATUS);
 	msg_notify.extension = 0;
 	msg_notify.tx_size = 0;
+	list_init(&msg_notify.list);
```
Comment (lyakh) [inline src/ipc/ipc4/handler.c:989]: I see that it affects the test in `ipc_msg_send()` but why is that test needed at all? Why is it possible to call that function for a message, that is already queued?..

File: src/include/sof/ipc/common.h [hunk 2]
```diff
@@ -65,6 +65,9 @@ struct ipc {
 	/* processing task */
 	struct task ipc_task;
 
+	/* task for flushing outbound messages */
```
Comment (lyakh) [inline src/include/sof/ipc/common.h:68]: to the commit message: "the platform driver **returns** -EBUSY?"

**Review by lyakh — APPROVED** [review]
very nice! Great work! Just would be good to understand why and how can messages be submitted to the queue repeatedly... Perhaps the reason is that SOF IPC messages are persistent objects, and the same object can be sent again by the owner while it is still in processing from the previous sending attempt and so it's on the queue. But maybe then owners should be fixed to not do that...

File: src/ipc/ipc4/handler.c (hunk 1 above)
Comment (ujfalusi) [reply src/ipc/ipc4/handler.c:989]: Is it OK to list_init() every time?
The problem is that we have single, local `msg_notify` and `msg_reply`.
If for example the notify got delayed, put to the list and you will have another notification: you change that message which was queued, list_init it again and things will be in an interesting state, no?

File: src/include/sof/ipc/common.h (hunk 2 above)
Comment (kv2019i) [reply src/include/sof/ipc/common.h:68]: Thanks, fixed in #6243

File: src/ipc/ipc4/handler.c (hunk 1 above)
Comment (kv2019i) [reply src/ipc/ipc4/handler.c:989]: @lyakh We might have contention between FW->HOST replies and FW->HOST notifications. Kernel needs to ack both and until kernel ack is received, ipc_msg_send() will return EBUSY.

File: src/drivers/intel/cavs/ipc.c [hunk 3]
```diff
@@ -185,6 +194,15 @@ enum task_state ipc_platform_do_cmd(struct ipc *ipc)
...
 
+#ifdef IPC4_EARLY_ACK
+	/*
+	 * notify host that we have received the message and
+	 * do this before we send the response
+	 */
+	ipc_write(IPC_DIPCTDR, ipc_read(IPC_DIPCTDR) | IPC_DIPCTDR_BUSY);
+	ipc_write(IPC_DIPCTDA, ipc_read(IPC_DIPCTDA) | IPC_DIPCTDA_DONE);
+#endif
+
```
Comment (RanderWang) [inline src/drivers/intel/cavs/ipc.c:205]: It is a test. I prefer to do it in ipc4/handler.c without ifdef.

Comment (kv2019i) [reply src/drivers/intel/cavs/ipc.c:205]: @RanderWang  Not sure I follow. This is writing to IPC hw registers and it thus should be done in the driver, not the protocol code in ipc4/handler.c. Right? The ifdef is needed as same driver is used for IPC3 and there the semantics are different ACK/REPLY are the same thing, so the proper place to do is later.

File: src/ipc/ipc4/handler.c (hunk 1 above)
Comment (lyakh) [reply src/ipc/ipc4/handler.c:1021]: @kv2019i Just for the record - you probably mean `ipc_platform_msg_send()`. Ok, so apparently either there's no waiting for the host response or the wait can be preempted.

**Comment by kv2019i** [comment]
Closing this at least for now. #6239 won't be merged in its current form.

**Pull request closed without merging**